pennylane_lightning/src/gates/cpu_kernels/avx_common/ApplyCRX.hpp

```diff
@@ -0,0 +1,289 @@
+        const size_t max_wire_parity = fillTrailingOnes(control);
+        const size_t max_wire_parity_inv = fillLeadingOnes(control + 1);
+
+        constexpr static auto perm = applyExternalInternalOffDiagPerm<target>();
+
+        if (inverse) {
+            angle *= -1.0;
+        }
+        const auto diag_factor =
+            set1<PrecisionT, packed_size>(std::cos(angle / 2));
+        const auto offdiag_factor =
+            imagFactor<PrecisionT, packed_size>(-std::sin(angle / 2));
+
+        for (size_t k = 0; k < exp2(num_qubits - 1); k += packed_size / 2) {
+            const size_t i0 =
+                ((k << 1U) & max_wire_parity_inv) | (max_wire_parity & k);
+            const size_t i1 = i0 | control_shift;
+
+            const auto v1 =
+                PrecisionAVXConcept::load(arr + i1); // control bit is 1
+            const auto w1 = Permutation::permute<perm>(v1);
+            PrecisionAVXConcept::store(arr + i1,
+                                       diag_factor * v1 + offdiag_factor * w1);
+        }
+    }
+
+    template <typename ParamT>
+    static void applyExternalExternal(std::complex<PrecisionT> *arr,
+                                      const size_t num_qubits,
+                                      const size_t control, const size_t target,
+                                      bool inverse, ParamT angle) {
+        using namespace Permutation;
+        const size_t control_shift = static_cast<size_t>(1U) << control;
+        const size_t target_shift = static_cast<size_t>(1U) << target;
+
+        const size_t rev_wire_min = std::min(control, target);
+        const size_t rev_wire_max = std::max(control, target);
+
+        const size_t parity_low = fillTrailingOnes(rev_wire_min);
+        const size_t parity_high = fillLeadingOnes(rev_wire_max + 1);
+        const size_t parity_middle =
+            fillLeadingOnes(rev_wire_min + 1) & fillTrailingOnes(rev_wire_max);
+
+        if (inverse) {
+            angle *= -1.0;
+        }
+
+        const auto cos_factor =
+            set1<PrecisionT, packed_size>(std::cos(angle / 2));
+        const auto sin_factor =
+            imagFactor<PrecisionT, packed_size>(-std::sin(angle / 2));
+
+        constexpr static auto perm = compilePermutation<PrecisionT>(
+            swapRealImag(identity<packed_size>()));
+
+        for (size_t k = 0; k < exp2(num_qubits - 2); k += packed_size / 2) {
+            const size_t i00 = ((k << 2U) & parity_high) |
+                               ((k << 1U) & parity_middle) | (k & parity_low);
+            const size_t i10 = i00 | control_shift;
+            const size_t i11 = i00 | control_shift | target_shift;
+
+            const auto v10 = PrecisionAVXConcept::load(arr + i10); // 10
+            const auto v11 = PrecisionAVXConcept::load(arr + i11); // 11
+
+            PrecisionAVXConcept::store(
+                arr + i10, cos_factor * v10 + sin_factor * permute<perm>(v11));
```

Comment:
I'm wondering if we can use 'PrecisionAVXConcept::mul' instead of '*' operator here?

Reply:
Here I use "*" defined in GCC vector extension (https://gcc.gnu.org/onlinedocs/gcc/Vector-Extensions.html). One advantage of using "*" and "+" directly is that the compiler optimizes pluses and products to FMA instructions. I want to use compiler-optimized FMA instructions instead of my hard-coded one, as it is easier to read (and maybe more optimal than what I may write).

Reply:
Yea, I agree.